config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java

```diff
@@ -0,0 +1,95 @@
+package com.typesafe.config.impl;
+
+import com.typesafe.config.ConfigNode;
+
+import java.util.*;
+
+final class ConfigNodeComplexValue implements ConfigNode, ConfigNodeValue {
+    private ArrayList<ConfigNode> children;
+    final private LinkedHashMap<Path, Integer> map = new LinkedHashMap<>();
+    final ArrayList<Integer> keyValueIndexes;
+
+    ConfigNodeComplexValue(Collection<ConfigNode> children) {
+        this.children = new ArrayList(children);
+        keyValueIndexes = new ArrayList<Integer>();
+
+        // Construct the list of indexes of Key-Value nodes. Do this
+        // in reverse order, since all but the final duplicate will be removed.
+        for (int i = this.children.size() - 1; i >= 0; i--) {
+            ConfigNode currNode = this.children.get(i);
+            if (currNode instanceof ConfigNodeKeyValue) {
+                keyValueIndexes.add(i);
+            }
+        }
```

Comment:
I'd probably say skip this; it's precomputing an index that's only used in `changeValueOnPath` (correct?) but for most nodes we will never change a value, so we're permanently using memory and doing CPU work up-front that often won't be necessary. We could compute it on the fly only when we actually change a value.